File: ios/Foodfolio/Core/API/DTO/RecipeDTO.swift
import Foundation

struct IngredientDTO: Codable, Equatable, Sendable {
  let id: String
  let name: String
  let amount: String?
  let sortOrder: Int
}
struct RecipeStepDTO: Codable, Equatable, Sendable {
  let id: String
  let text: String
  let sortOrder: Int
}
struct TagDTO: Codable, Equatable, Sendable {
  let id: String
  let name: String
  let createdAt: Date
}
struct RecipeDTO: Codable, Equatable, Sendable {
  let id: String
  let originalUrl: String
  let sourceType: String
  let title: String
  let imageUrl: String?
  let servingsValue: Double?
  let servingsRaw: String?
  let cookingTimeMinutes: Int?
  let genre: String?
  let analysisStatus: AnalysisStatus
  let wantToCookAt: Date?
  let ingredients: [IngredientDTO]
  let steps: [RecipeStepDTO]
  let tags: [TagDTO]
  let createdAt: Date
  let updatedAt: Date

  init(
    id: String, originalUrl: String, sourceType: String, title: String, imageUrl: String?,
    servingsValue: Double?, servingsRaw: String?, cookingTimeMinutes: Int?, genre: String?,
    analysisStatus: AnalysisStatus, wantToCookAt: Date? = nil, ingredients: [IngredientDTO],
    steps: [RecipeStepDTO], tags: [TagDTO], createdAt: Date, updatedAt: Date
  ) {
    self.id = id
    self.originalUrl = originalUrl
    self.sourceType = sourceType
    self.title = title
    self.imageUrl = imageUrl
    self.servingsValue = servingsValue
    self.servingsRaw = servingsRaw
    self.cookingTimeMinutes = cookingTimeMinutes
    self.genre = genre
    self.analysisStatus = analysisStatus
    self.wantToCookAt = wantToCookAt
    self.ingredients = ingredients
    self.steps = steps
    self.tags = tags
    self.createdAt = createdAt
    self.updatedAt = updatedAt
  }
}
struct SyncResponse: Codable, Sendable {
  let recipes: [RecipeDTO]
  let tags: [TagDTO]
  let nextCursor: String
}
struct RecipeIDsResponse: Codable, Sendable { let recipeIds: [String] }
struct SettingDTO: Codable, Sendable { let recipeAnalysisNotificationEnabled: Bool }

File: src/api/recipe-dto.ts
import type { Prisma } from "../generated/prisma/client.js";
import { GENRE_LABELS } from "../domain/recipe/genre.js";

export const recipeInclude = {
  ingredients: { orderBy: { sortOrder: "asc" as const } },
  steps: { orderBy: { sortOrder: "asc" as const } },
  recipeTags: { include: { tag: true } },
} satisfies Prisma.RecipeInclude;

export type FullRecipe = Prisma.RecipeGetPayload<{
  include: typeof recipeInclude;
}>;

export function recipeDto(recipe: FullRecipe) {
  return {
    id: recipe.id,
    originalUrl: recipe.originalUrl,
    sourceType: recipe.sourceType,
    title: recipe.title,
    imageUrl: recipe.imageUrl,
    servingsValue: recipe.servingsValue,
    servingsRaw: recipe.servingsRaw,
    cookingTimeMinutes: recipe.cookingTimeMinutes,
    genre: recipe.genre ? GENRE_LABELS[recipe.genre] : null,
    analysisStatus: recipe.analysisStatus,
    wantToCookAt: recipe.wantToCookAt?.toISOString() ?? null,
    ingredients: recipe.ingredients.map(({ id, name, amount, sortOrder }) => ({
      id,
      name,
      amount,
      sortOrder,
    })),
    steps: recipe.steps.map(({ id, text, sortOrder }) => ({
      id,
      text,
      sortOrder,
    })),
    tags: recipe.recipeTags.map(({ tag }) => ({
      id: tag.id,
      name: tag.name,
      createdAt: tag.createdAt.toISOString(),
    })),
    createdAt: recipe.createdAt.toISOString(),
    updatedAt: recipe.updatedAt.toISOString(),
  };
}
